### PR TITLE
protocol: more restrictive checks on chain entry response

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

**Review by selsta — APPROVED** [review]

File: src/cryptonote_protocol/cryptonote_protocol_handler.inl [hunk 1]
```diff
@@ -2532,14 +2532,21 @@ skip:
...
     for (size_t i = 0; i < arg.m_block_ids.size(); ++i)
     {
+      blocks_found.insert(arg.m_block_ids[i]);
       const uint64_t block_weight = arg.m_block_weights.empty() ? 0 : arg.m_block_weights[i];
       context.m_needed_objects.push_back(std::make_pair(arg.m_block_ids[i], block_weight));
       if (++added == n_use_blocks)
         break;
     }
-    context.m_last_response_height -= arg.m_block_ids.size() - n_use_blocks;
+    if (blocks_found.size() != arg.m_block_ids.size())
```
Comment (SChernykh) [inline src/cryptonote_protocol/cryptonote_protocol_handler.inl:2544]: This assumes that the previous loop always iterates without early exit. Are you sure `n_use_blocks` is always the same as `arg.m_block_ids.size()`?

File: src/cryptonote_protocol/cryptonote_protocol_handler.inl [hunk 2]
```diff
@@ -2486,7 +2486,7 @@ skip:
       drop_connection(context, true, false);
       return 1;
     }
-    if (arg.total_height < arg.m_block_ids.size() || arg.start_height > arg.total_height - arg.m_block_ids.size())
+    if (arg.total_height < arg.m_block_ids.size() || arg.start_height > arg.total_height - arg.m_block_ids.size() || arg.start_height >= m_core.get_current_blockchain_height())
```
Comment (SChernykh) [inline src/cryptonote_protocol/cryptonote_protocol_handler.inl:2489]: Just to clarify. Is it a response to our request, so starting height must always be below current blockchain height?

File: src/cryptonote_protocol/cryptonote_protocol_handler.inl (hunk 1 above)
Comment (SChernykh) [reply src/cryptonote_protocol/cryptonote_protocol_handler.inl:2544]: I think it's safer to check the return value of `blocks_found.insert(arg.m_block_ids[i])` to see if it actually inserted new element or not. From C++ docs:
>  the function returns a pair object whose first element is an iterator pointing either to the newly inserted element in the container or to the element whose key is equivalent, and a bool value indicating whether the element was successfully inserted or not.

File: src/cryptonote_protocol/cryptonote_protocol_handler.inl (hunk 2 above)
Comment (moneromooo-monero) [reply src/cryptonote_protocol/cryptonote_protocol_handler.inl:2489]: It is.

**Review by SChernykh — APPROVED** [review]

**Pull request merged**